### PR TITLE
add RouteTable api

### PR DESCRIPTION
**Pull request by ilackarms**
adds the new RouteTable api intended for use with delegation

related: #1062 
BOT NOTES: 
resolves https://github.com/solo-io/gloo/pull/1111

### CODE REVIEW

File: install/helm/gloo/templates/100-gloo-crds.yaml [hunk 1]
```diff
@@ -68,6 +68,28 @@ spec:
...
+  annotations:
+    "helm.sh/hook": crd-install
+spec:
+  group: gateway.solo.io
+  names:
+    kind: RouteTable
+    listKind: RouteTableList
+    plural: routetables
+    shortNames:
+      - vs
```
Comment (kdorosh) [inline install/helm/gloo/templates/100-gloo-crds.yaml]: this should be `rt`?

File: install/helm/gloo/templates/100-gloo-crds.yaml [hunk 2]
```diff
@@ -68,6 +68,28 @@ spec:
...
+metadata:
+  name: routetables.gateway.solo.io
+  annotations:
+    "helm.sh/hook": crd-install
+spec:
+  group: gateway.solo.io
+  names:
+    kind: RouteTable
+    listKind: RouteTableList
+    plural: routetables
```
Comment (kdorosh) [inline install/helm/gloo/templates/100-gloo-crds.yaml]: this doesn't match the plural `route_tables` referenced in the proto

Comment (kdorosh) [reply install/helm/gloo/templates/100-gloo-crds.yaml]: make sure matches https://github.com/solo-io/gloo/pull/1112/files#diff-bf23b90d1e4a3b0665e720552691c8aeR67-R68

File: install/helm/gloo/templates/100-gloo-crds.yaml (hunk 1 above)
Comment (kdorosh) [reply install/helm/gloo/templates/100-gloo-crds.yaml]: make sure matches https://github.com/solo-io/gloo/pull/1112/files#diff-bf23b90d1e4a3b0665e720552691c8aeR67-R68

File: install/helm/gloo/templates/100-gloo-crds.yaml (hunk 2 above)
Comment (ilackarms) [reply install/helm/gloo/templates/100-gloo-crds.yaml]: it's the same with `virtual_services`. cant use underscores in the kube plural name

File: projects/gateway/api/v1/virtual_service.proto [hunk 3]
```diff
@@ -7,27 +7,122 @@ option (gogoproto.equal_all) = true;
...
+
 */
 message VirtualService {
 
     option (core.solo.io.resource).short_name = "vs";
     option (core.solo.io.resource).plural_name = "virtual_services";
 
-    gloo.solo.io.VirtualHost virtual_host = 1;
+    // The VirtualHost contains the
+    // The list of HTTP routes define routing actions to be taken for incoming HTTP requests whose host header matches
```
Comment (kdorosh) [inline projects/gateway/api/v1/virtual_service.proto]: nit: newline/spacing here is awkward

File: projects/gateway/api/v1/virtual_service.proto [hunk 4]
```diff
@@ -36,4 +131,194 @@ message VirtualService {
...
+        // Return an arbitrary HTTP response directly, without proxying.
+        gloo.solo.io.DirectResponseAction direct_response_action = 4;
+
+        // delegate routing actions for the given matcher to a RouteTable
+        // the delegateAction config is simply the `name` and `namespace` of the delegated `RouteTable` resource
+        core.solo.io.ResourceRef delegate_action = 5;
+    }
+
+    // Route Plugins extend the behavior of routes.
+    // Route plugins include configuration such as retries,rate limiting, and request/response transformation.
```
Comment (kdorosh) [inline projects/gateway/api/v1/virtual_service.proto]: nit: typo
```suggestion
    // Route plugins include configuration such as retries, rate limiting, and request/response transformation.
```

File: projects/gateway/api/v2/gateway.proto [hunk 5]
```diff
@@ -54,13 +55,26 @@ message Gateway {
         TcpGateway tcp_gateway = 10;
     }
 
+    // deprecated: use proxyNames
```
Comment (kdorosh) [inline projects/gateway/api/v2/gateway.proto]: 👏

File: projects/gateway/pkg/translator/http.go [hunk 6]
```diff
@@ -210,3 +208,46 @@ func desiredListenerForHttp(gateway *v2.Gateway, virtualServicesForGateway v1.Vi
...
+	case *v1.Route_DirectResponseAction:
+		route.Action = &gloov1.Route_RedirectAction{
+			RedirectAction: ours.GetRedirectAction(),
+		}
+	case *v1.Route_RouteAction:
+		route.Action = &gloov1.Route_RedirectAction{
+			RedirectAction: ours.GetRedirectAction(),
+		}
+	case *v1.Route_DelegateAction:
+		panic("not implemented")
```
Comment (kdorosh) [inline projects/gateway/pkg/translator/http.go]: is there a separate issue open for this?

Comment (ilackarms) [reply projects/gateway/pkg/translator/http.go]: #1062 is tracking the feature

File: projects/gateway/api/v2/gateway.proto [hunk 7]
```diff
@@ -54,13 +55,26 @@ message Gateway {
...
 }
 
 message HttpGateway {
     // names of the the virtual services, which contain the actual routes for the gateway
     // if the list is empty, all virtual services will apply to this gateway (with accordance to tls flag above).
     repeated core.solo.io.ResourceRef virtual_services = 1 [(gogoproto.nullable) = false];
+
+    // only one of `virtualServices` or `virtualServiceSelector` should be provided
+    // rather than providing
+    map<string, string> virtual_service_selector = 2 ;
```
Comment (EItanya) [inline projects/gateway/api/v2/gateway.proto]: use a proto oneof?

File: projects/gateway/api/v2/gateway.proto [hunk 8]
```diff
@@ -54,13 +55,26 @@ message Gateway {
...
 
+    // deprecated: use proxyNames
     string gateway_proxy_name = 11;
+
+    // By default, each Gateway is exposed on the default `gateway-proxy-v2` which lives in the Gloo installation namespace.
+    // Use `proxyNames` to expose Gateways on custom proxies deployed by the user via the [Gloo Helm Chart](https://gloo.solo.io/installation/gateway/kubernetes/#list-of-gloo-helm-chart-values)
+    //
+    // > Note: when using camel-case proxy names in the Helm chart, they should be converted to kebab-case for use with `proxyNames`, such that:
+    // `gatewayProxyV2` in the Helm Values becomes `gateway-proxy-v2` in the list of proxyNames
+    string proxy_names = 12;
```
Comment (kdorosh) [inline projects/gateway/api/v2/gateway.proto]: this should be a repeated string

Comment (EItanya) [inline projects/gateway/api/v2/gateway.proto]: repeated string

Comment (kdorosh) [reply projects/gateway/api/v2/gateway.proto]: > they should be converted to kebab-case for use with `proxyNames`, such that:
>    // `gatewayProxyV2` in the Helm Values becomes `gateway-proxy-v2` in the list of proxyNames

This is leaking implementation details. We should either derive it from the helm chart or do the case conversion for them

Comment (ilackarms) [reply projects/gateway/api/v2/gateway.proto]: my idea was to decouple from the helm chart. i could reword this comment to read

> Name of the proxy resource to generate from this gateway. If other gateways exist which point to the same proxy, Gloo will join them together.
> Defaults to `gateway-proxy-v2`
> Note: this should map to the [`node.metadata.role`](https://github.com/solo-io/gloo/blob/master/install/helm/gloo/templates/9-gateway-proxy-configmap.yaml#L22) key in the Envoy bootstrap config.

File: projects/gateway/api/v2/gateway.proto (hunk 7 above)
Comment (ilackarms) [reply projects/gateway/api/v2/gateway.proto]: unfortunately, [protobuf doesn't support putting repeated fields in oneofs](https://github.com/protocolbuffers/protobuf/issues/2592). to use a oneof, we'd need to create a wrapper object and we'd get a breaking change. 😢

**Comment by ilackarms** [comment]
/kick

```

Step #12 - "docker-login": WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Step #12 - "docker-login": Error response from daemon: Get https://quay.io
```

**Pull request merged**